### PR TITLE
fix: migrate showConstraintsQuery to TypeScript

### PR DESCRIPTION
**Pull request by lohart13**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change
Add missing showConstraint query typing.
Migrate showConstraintsQuery to TypeScript

~~The results for the showConstraint seems to differ from dialect to dialect so I've made the constraint definition fairly basic.
Is this something that needs to be unified, so each dialect returns the same data?~~

## Todos

- [x] Check out the constraint definition by dialect

### CODE REVIEW

**Review by WikiRik — COMMENTED** [review]
Looking at https://github.com/sequelize/sequelize/blob/main/test/unit/sql/show-constraints.test.js it seems that it might return a lot more than that. I'm not sure which of these would be useful and I think we need to look into the entire constraints set of functions since that could use some updating and unifying of tests. Maybe leave this small improvement for a bigger PR?

**Comment by lohart13** [comment]
No worries, happy to close in view of a larger PR.
Do you want me to raise an issue instead?

**Comment by WikiRik** [comment]
That might be nice, that way we won't forget about it. Thanks!

**Comment by lohart13** [comment]
I've updated this PR and moved the `showConstraintQuery` function to typescript.

LMK, if you have any questions

File: src/dialects/abstract/query-interface.d.ts [hunk 1]
```diff
@@ -274,6 +274,25 @@ export interface IndexDescription {
   type: string | undefined;
 }
 
+export interface ConstraintDescription {
+  constrainCatalog: string | null | undefined;
```
Comment (WikiRik) [inline src/dialects/abstract/query-interface.d.ts:278]:
```suggestion
  constraintCatalog?: string;
```
Both a typo and a suggestion. If we really want it to allow `null` as well you can add that but I'd prefer not to. The suggestion holds for the other properties in this interface as well

File: src/dialects/db2/query-generator-typescript.ts [hunk 2]
```diff
@@ -27,6 +27,33 @@ export class Db2QueryGeneratorTypeScript extends AbstractQueryGenerator {
...
+      `CASE r.DELETERULE WHEN 'A' THEN 'NO ACTION' WHEN 'C' THEN 'CASCADE' WHEN 'N' THEN 'SET NULL' WHEN 'R' THEN 'RESTRICT' ELSE NULL END AS "deleteRule",`,
+      `CASE r.UPDATERULE WHEN 'A' THEN 'NO ACTION' WHEN 'R' THEN 'RESTRICT' ELSE NULL END AS "updateRule",`,
+      'ck.TEXT AS "definition"',
+      'FROM SYSCAT.TABCONST c',
+      'LEFT JOIN SYSCAT.REFERENCES r ON c.CONSTNAME = r.CONSTNAME AND c.TABNAME = r.TABNAME AND c.TABSCHEMA = r.TABSCHEMA',
+      'LEFT JOIN SYSCAT.KEYCOLUSE k ON r.CONSTNAME = k.CONSTNAME AND r.TABNAME = k.TABNAME AND r.TABSCHEMA = k.TABSCHEMA',
+      'LEFT JOIN SYSCAT.KEYCOLUSE fk ON r.REFKEYNAME = fk.CONSTNAME',
+      'LEFT JOIN SYSCAT.CHECKS ck ON c.CONSTNAME = ck.CONSTNAME AND c.TABNAME = ck.TABNAME AND c.TABSCHEMA = ck.TABSCHEMA',
+      `WHERE c.TABNAME = ${this.escape(table.tableName)}`,
+      table.schema !== '' ? `AND c.TABSCHEMA = ${this.escape(table.schema)}` : 'AND c.TABSCHEMA = USER',
```
Comment (WikiRik) [inline src/dialects/db2/query-generator-typescript.ts:51]:
```suggestion
      'AND c.TABSCHEMA',
      table.schema ? this.escape(table.schema) : 'USER',
```
This is in line with the changes I proposed in #15454

File: src/dialects/ibmi/query-generator-typescript.ts [hunk 3]
```diff
@@ -33,6 +33,37 @@ export class IBMiQueryGeneratorTypeScript extends AbstractQueryGenerator {
...
+      'c.IS_DEFERRABLE AS "isDeferrable",',
+      'c.INITIALLY_DEFERRED AS "initiallyDeferred"',
+      'FROM QSYS2.SYSCST c',
+      'LEFT JOIN QSYS2.SYSREFCST r ON c.CONSTRAINT_NAME = r.CONSTRAINT_NAME AND c.CONSTRAINT_SCHEMA = r.CONSTRAINT_SCHEMA',
+      'LEFT JOIN QSYS2.SYSKEYCST k ON r.CONSTRAINT_NAME = k.CONSTRAINT_NAME AND r.CONSTRAINT_SCHEMA = k.CONSTRAINT_SCHEMA',
+      'LEFT JOIN QSYS2.SYSKEYCST fk ON r.UNIQUE_CONSTRAINT_NAME = k.CONSTRAINT_NAME AND r.UNIQUE_CONSTRAINT_SCHEMA = k.CONSTRAINT_SCHEMA',
+      'LEFT JOIN QSYS2.SYSCHKCST ch ON c.CONSTRAINT_NAME = ch.CONSTRAINT_NAME AND c.CONSTRAINT_SCHEMA = ch.CONSTRAINT_SCHEMA',
+      `WHERE c.TABLE_NAME = ${this.escape(table.tableName)}`,
+      'AND c.TABLE_SCHEMA =',
+      table.schema !== '' ? this.escape(table.schema) : 'CURRENT SCHEMA',
```
Comment (WikiRik) [inline src/dialects/ibmi/query-generator-typescript.ts:61]:
```suggestion
      table.schema ? this.escape(table.schema) : 'CURRENT SCHEMA',
```

File: src/dialects/mysql/query-generator-typescript.ts [hunk 4]
```diff
@@ -14,6 +16,63 @@ export class MySqlQueryGeneratorTypeScript extends AbstractQueryGenerator {
     return `SHOW FULL COLUMNS FROM ${this.quoteTable(tableName)};`;
   }
 
+  showConstraintsQuery(tableName: TableNameOrModel, constraintName?: string) {
+    const table = this.extractTableDetails(tableName);
+    const dbVersion = this.sequelize.getDatabaseVersion() || '0.0.0';
+
+    // MySQL 8.0.16+ has a new INFORMATION_SCHEMA.CHECK_CONSTRAINTS table
```
Comment (WikiRik) [inline src/dialects/mysql/query-generator-typescript.ts:23]: Looking at the integration tests it seems that we have a different behaviour depending on the used databaseVersion. Would it be possible to have the same behaviour on MySQL 5.7 and 8.0.0-8.0.15? I'd prefer not to have different queries depending on the databaseVersion

File: src/dialects/mysql/query-generator-typescript.ts [hunk 5]
```diff
@@ -14,6 +16,63 @@ export class MySqlQueryGeneratorTypeScript extends AbstractQueryGenerator {
...
+        'ch.CHECK_CLAUSE AS definition',
+        'FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS c',
+        'LEFT JOIN INFORMATION_SCHEMA.REFERENTIAL_CONSTRAINTS r ON c.CONSTRAINT_CATALOG = r.CONSTRAINT_CATALOG',
+        'AND c.CONSTRAINT_SCHEMA = r.CONSTRAINT_SCHEMA AND c.CONSTRAINT_NAME = r.CONSTRAINT_NAME AND c.TABLE_NAME = r.TABLE_NAME',
+        'LEFT JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu ON r.CONSTRAINT_CATALOG = kcu.CONSTRAINT_CATALOG',
+        'AND r.CONSTRAINT_SCHEMA = kcu.CONSTRAINT_SCHEMA AND r.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME AND r.TABLE_NAME = kcu.TABLE_NAME',
+        'LEFT JOIN INFORMATION_SCHEMA.CHECK_CONSTRAINTS ch ON c.CONSTRAINT_CATALOG = ch.CONSTRAINT_CATALOG',
+        'AND c.CONSTRAINT_SCHEMA = ch.CONSTRAINT_SCHEMA AND c.CONSTRAINT_NAME = ch.CONSTRAINT_NAME',
+        `WHERE c.TABLE_NAME = ${this.escape(table.tableName)}`,
+        table.schema !== '' ? `AND c.TABLE_SCHEMA = ${this.escape(table.schema)}` : '',
```
Comment (WikiRik) [inline src/dialects/mysql/query-generator-typescript.ts:46]:
```suggestion
        `AND c.TABLE_SCHEMA = ${this.escape(table.schema)}`,
```
Apart from db2 and ibmi we always return a schema

File: src/dialects/mysql/query-generator-typescript.ts [hunk 6]
```diff
@@ -14,6 +16,63 @@ export class MySqlQueryGeneratorTypeScript extends AbstractQueryGenerator {
...
+      'kcu.REFERENCED_COLUMN_NAME AS referencedColumnName,',
+      'r.DELETE_RULE AS deleteAction,',
+      'r.UPDATE_RULE AS updateAction',
+      'FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS c',
+      'LEFT JOIN INFORMATION_SCHEMA.REFERENTIAL_CONSTRAINTS r ON c.CONSTRAINT_CATALOG = r.CONSTRAINT_CATALOG',
+      'AND c.CONSTRAINT_SCHEMA = r.CONSTRAINT_SCHEMA AND c.CONSTRAINT_NAME = r.CONSTRAINT_NAME AND c.TABLE_NAME = r.TABLE_NAME',
+      'LEFT JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu ON r.CONSTRAINT_CATALOG = kcu.CONSTRAINT_CATALOG',
+      'AND r.CONSTRAINT_SCHEMA = kcu.CONSTRAINT_SCHEMA AND r.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME AND r.TABLE_NAME = kcu.TABLE_NAME',
+      `WHERE c.TABLE_NAME = ${this.escape(table.tableName)}`,
+      table.schema !== '' ? `AND c.TABLE_SCHEMA = ${this.escape(table.schema)}` : '',
```
Comment (WikiRik) [inline src/dialects/mysql/query-generator-typescript.ts:70]:
```suggestion
       `AND c.TABLE_SCHEMA = ${this.escape(table.schema)}`,
```

File: src/dialects/sqlite/query-generator.js [hunk 7]
```diff
@@ -366,18 +366,10 @@ export class SqliteQueryGenerator extends SqliteQueryGeneratorTypeScript {
     return result;
   }
 
-  showConstraintsQuery(tableName, constraintName) {
-    let sql = `SELECT sql FROM sqlite_master WHERE tbl_name='${tableName}'`;
-
-    if (constraintName) {
-      sql += ` AND sql LIKE '%${constraintName}%'`;
```
Comment (WikiRik) [inline src/dialects/sqlite/query-generator.js:373]: Why was this removed? Did it not work as expected?

File: src/dialects/sqlite/query-generator-typescript.ts [hunk 8]
```diff
@@ -15,6 +15,16 @@ export class SqliteQueryGeneratorTypeScript extends MySqlQueryGenerator {
     return `PRAGMA TABLE_INFO(${this.quoteTable(tableName)});`;
   }
 
+  showConstraintsQuery(tableName: TableNameOrModel, _constraintName?: string) {
+    const table = this.extractTableDetails(tableName);
+
+    return joinSQLFragments([
+      'SELECT sql FROM sqlite_master',
+      `WHERE tbl_name = ${this.escape(table.tableName)}`,
+      ';',
```
Comment (WikiRik) [inline src/dialects/sqlite/query-generator-typescript.ts:24]:
```suggestion
```
I'm removing the trailing semicolon from here, see #15454

File: src/dialects/sqlite/query-generator.js [hunk 9]
```diff
@@ -366,18 +366,10 @@ export class SqliteQueryGenerator extends SqliteQueryGeneratorTypeScript {
...
-    }
-
-    return `${sql};`;
-  }
-
   describeCreateTableQuery(tableName) {
```
Comment (WikiRik) [inline src/dialects/sqlite/query-generator.js:369]: Can you add the typings for this as well?

**Review by WikiRik — COMMENTED** [review]
I'll need to take a further look at the new queries and tests, but these are my initial comments. Great work!

File: src/dialects/sqlite/query-generator-typescript.ts (hunk 8 above)
Comment (lohart13) [reply src/dialects/sqlite/query-generator-typescript.ts]: Done for all dialects

File: src/dialects/sqlite/query-generator.js (hunk 9 above)
Comment (lohart13) [reply src/dialects/sqlite/query-generator.js:369]: Done

File: src/dialects/abstract/query-generator.js [hunk 10]
```diff
@@ -703,6 +703,7 @@ export class AbstractQueryGenerator extends AbstractQueryGeneratorTypeScript {
   }
 
   getConstraintSnippet(tableName, options) {
```
[the remark this answers is not in the document]
Comment (lohart13) [reply src/dialects/abstract/query-generator.js:705]: Done

File: src/dialects/abstract/query-interface.d.ts (hunk 1 above)
Comment (lohart13) [reply src/dialects/abstract/query-interface.d.ts]: I've updated the typing now, let me know if you're happy with this.

File: src/dialects/mysql/query-generator-typescript.ts (hunk 4 above)
Comment (lohart13) [reply src/dialects/mysql/query-generator-typescript.ts:23]: The issue is that in MySQL version 8.0.16+ has a new INFORMATION_SCHEMA.CHECK_CONSTRAINTS table which allows the `showConstraintsQuery` to list the check constraints however in versions prior, this is not possible and the query will return an error.

File: src/dialects/sqlite/query-generator.js (hunk 7 above)
Comment (lohart13) [reply src/dialects/sqlite/query-generator.js]: I can add this back in however it's passing a name is not going to filter the results to the `constraintName` value, all constraints will be returned for the table unless the value of `constraintName` doesn't exist in which case an empty array would be returned.

It might be better to throw an error if a value is passed informing the user that SQLite doesn't support filtering constraints.

**Comment by lohart13** [comment]
I'm not sure why the tests have failed, from my local testing the test pass

File: packages/core/test/unit/query-generator/show-constraints-query.test.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+      sqlite: `SELECT sql FROM sqlite_master WHERE tbl_name = 'myTable'`,
+    });
+  });
+
+  it('produces a show constraints query from a model', () => {
+    const MyModel = sequelize.define('myModel', {});
```
Comment (WikiRik) [inline packages/core/test/unit/query-generator/show-constraints-query.test.ts:34]:
```suggestion
    const MyModel = sequelize.define('MyModel', {});
```

Also needs to be updated in the tests themselves

**Review by WikiRik — COMMENTED** [review]
I will have to take a close look at this, but these were some initial comments I still have on pending. Most relate to making the unit test suite more in line with the other files in that folder

**Comment by lohart13** [comment]
@WikiRik for the MySQL comments, support for check constraints was only added in version 8.0.16 as per https://dev.mysql.com/blog-archive/mysql-8-0-16-introducing-check-constraint

Do you want to enable support for check constraints for MySQL at all?
If so, I can update the PR to throw an error if a check constraint is added when the version lower than 8.0.16.
Otherwise, I'll revert the MySQL changes.

File: packages/core/test/unit/query-generator/show-constraints-query.test.ts (hunk 11 above)
Comment (lohart13) [reply packages/core/test/unit/query-generator/show-constraints-query.test.ts]: Done

File: packages/core/test/unit/query-generator/show-constraints-query.test.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
+import { createSequelizeInstance, expectsql } from '../../support';
+
+const sequelize = createSequelizeInstance({ databaseVersion: '0.0.0' });
+const dialect = sequelize.dialect;
+
+describe('QueryGenerator#showConstraintsQuery', () => {
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/show-constraints-query.test.ts:6]: Done

File: packages/core/test/integration/query-interface.test.js [hunk 13]
```diff
@@ -585,11 +586,15 @@ describe(Support.getTestDialectTeaser('QueryInterface'), () => {
...
           let constraints = await this.queryInterface.showConstraint('users');
           constraints = constraints.map(constraint => constraint.constraintName);
-          expect(constraints).to.include('check_user_roles');
-          await this.queryInterface.removeConstraint('users', 'check_user_roles');
-          constraints = await this.queryInterface.showConstraint('users');
-          constraints = constraints.map(constraint => constraint.constraintName);
-          expect(constraints).to.not.include('check_user_roles');
+          if (dialectName === 'mysql' && semver.lt(current.getDatabaseVersion(), '8.0.16')) {
+            // MySQL 8.0.16+ has a new INFORMATION_SCHEMA.CHECK_CONSTRAINTS table
+            expect(constraints).to.not.include('check_user_roles');
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/integration/query-interface.test.js:639]: I've added this back in

File: packages/core/test/integration/query-interface.test.js [hunk 14]
```diff
@@ -585,11 +586,15 @@ describe(Support.getTestDialectTeaser('QueryInterface'), () => {
           });
           let constraints = await this.queryInterface.showConstraint('users');
           constraints = constraints.map(constraint => constraint.constraintName);
-          expect(constraints).to.include('check_user_roles');
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/integration/query-interface.test.js:636]: I've added this line back in under the else statement so that MySQL version <8.0.16 test don't fail due to not supporting check constraints.

File: packages/core/test/integration/query-interface.test.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/integration/query-interface.test.js:639]: This will part of not be part of showConstraint if the dialect is MySQL and the version is < 8.0.16 as check constraints are not supported in older versions.
